src/transform/src/reduction_pushdown.rs

```diff
@@ -81,8 +90,707 @@ impl ReductionPushdown {
+    fn try_new(
+        group_key: &'a [MirScalarExpr],
+        aggregates: &[AggregateExpr],
+        inputs: &[MirRelationExpr],
+        equivalences: &'a [Vec<MirScalarExpr>],
+        implementation: &expr::JoinImplementation,
+        predicates: &[MirScalarExpr],
+    ) -> Option<Self> {
+        let input_types = inputs.iter().map(|i| i.typ()).collect::<Vec<_>>();
+        let old_join_mapper = JoinInputMapper::new_from_input_types(&input_types);
+
+        // sort aggregations by how and whether they can be pushed down
+        // TODO: attempt to rewrite multi-input aggregations in terms of a
+        // single input
+        // TODO: attempt to reduce the number of inputs that aggregations come
+        // from by rewriting aggregations on join keys in terms of the
+        // equivalent expression from the other input
+        let mut outer_aggs = Vec::new();
+        let mut single_input_aggs_to_push = vec![(Vec::new(), Vec::new()); inputs.len()];
+        let mut no_input_aggs_to_push = Vec::new();
+
+        // classify the aggregations by how many inputs each aggregation references
+        for (agg_num, agg) in aggregates.iter().enumerate() {
+            let mut expr_inputs = old_join_mapper.lookup_inputs(&agg.expr);
+            if agg.func.outer_agg().is_none() {
+                // Aggregates that cannot be partially pushed down go in the
+                // outer reduce.
+                outer_aggs.push((agg_num, agg.clone()));
+            } else if let AggregateFunc::Dummy = agg.func {
+                // Do not bother pushing dummy functions down.
+                outer_aggs.push((agg_num, agg.clone()));
+            } else if agg.distinct && agg.func.ignores_duplicates() {
+                outer_aggs.push((agg_num, agg.clone()));
+            } else if let Some(expr_input) = expr_inputs.next() {
+                if expr_inputs.next().is_some() {
+                    // an aggregation involving more than one input
+                    outer_aggs.push((agg_num, agg.clone()));
+                } else {
+                    // an aggregation involving one input.
+                    let mut localized_agg = agg.clone();
+                    localized_agg.expr = old_join_mapper.map_expr_to_local(localized_agg.expr);
+                    if let MirRelationExpr::Reduce { group_key, .. } = &inputs[expr_input] {
+                        if localized_agg
+                            .expr
+                            .support()
+                            .iter()
+                            .any(|i| i >= &group_key.len())
+                        {
+                            // skip pushing down the aggregation if it aggregates a
+                            // partial aggregate to prevent infinite layers of
+                            // partial reduction pushdown
+                            outer_aggs.push((agg_num, agg.clone()));
+                            continue;
+                        }
+                    }
```

Comment:
I'm not sure I follow the logic here. I understand not wanting to push down reductions repeatedly, but I can't follow the logic here. If it helps, the `reduction_elision` transformation will (should) remove a stack of reductions with the same key, so as long as it happens after this transformation, the net action should only be one reduction pushed down (that is, if this is complicated it can be deferred to a different transformation).

Reply:
I tried taking out the code above, and the `reduction_elision` transform does indeed remove any redundant reductions. 

But putting the `ReductionPushdown` back in the Fixpoint results in plans changing in funny ways because we haven't finished fusing multi-input joins by the time we reach the fix point, so reductions get pushed down multiple layers of joins and prevent the joins from being fused.

Example:
TPCH Q02 before:
```
<... everything up to %8 is the same>
%9 =
| Get materialize.public.nation (u1)
| ArrangeBy (#0) (#2)

%10 =
| Get materialize.public.region (u4)
| ArrangeBy (#0)

%11 =
| Join %6 %7 %8 %9 %10 (= #0 #1) (= #2 #6) (= #9 #13) (= #15 #17)
| | implementation = DeltaQuery
| |   delta %6 %7.(#0) %8.(#0) %9.(#0) %10.(#0)
| |   delta %7 %6.(#0) %8.(#0) %9.(#0) %10.(#0)
| |   delta %8 %9.(#0) %10.(#0) %7.(#1) %6.(#0)
| |   delta %9 %10.(#0) %8.(#3) %7.(#1) %6.(#0)
| |   delta %10 %9.(#2) %8.(#3) %7.(#1) %6.(#0)
| | demand = (#0, #4, #18)
| Filter (#18 = "EUROPE")
| Reduce group=(#0)
| | agg min(#4)
| ArrangeBy (#0, #1)

%12 =
| Join %5 %11 (= #0 #28) (= #19 #29)
| | implementation = Differential %5 %11.(#0, #1)
| | demand = (#0, #2, #10, #11, #13..#15, #22)
| Project (#14, #10, #22, #0, #2, #11, #13, #15)
```

TPCH Q02 after:
```
<... everything up to %8 is the same>

%9 =
| Join %6 %7 %8 (= #0 #1) (= #2 #6)
| | implementation = DeltaQuery
| |   delta %6 %7.(#0) %8.(#0)
| |   delta %7 %6.(#0) %8.(#0)
| |   delta %8 %7.(#1) %6.(#0)
| | demand = (#0, #4, #9)
| Reduce group=(#0, #9)
| | agg min(#4)

%10 =
| Get materialize.public.nation (u1)
| ArrangeBy (#0)

%11 =
| Join %9 %10 (= #1 #3)
| | implementation = Differential %9 %10.(#0)
| | demand = (#0, #2, #5)
| Reduce group=(#0, #5)
| | agg min(#2)

%12 =
| Get materialize.public.region (u4)
| ArrangeBy (#0)

%13 =
| Join %11 %12 (= #1 #3)
| | implementation = Differential %11 %12.(#0)
| | demand = (#0, #2, #4)
| Filter (#4 = "EUROPE")
| Reduce group=(#0)
| | agg min(#2)
| ArrangeBy (#0, #1)

%14 =
| Join %5 %13 (= #0 #28) (= #19 #29)
| | implementation = Differential %5 %13.(#0, #1)
| | demand = (#0, #2, #10, #11, #13..#15, #22)
| Project (#14, #10, #22, #0, #2, #11, #13, #15)

Finish order_by=(#0 desc, #2 asc, #1 asc, #3 asc) limit=none offset=0 project=(#0..#7)

EOF
```

On the one hand, it is not necessarily bad for there to be reductions in the intermediate state of a differential join, so I could do some performance testing of this. But instinctively, I don't like the new plan because we don't get to choose the join order.